src/python/pants/util/collections.py

```diff
@@ -63,3 +73,42 @@ def assert_single_element(iterable: Iterable[T]) -> T:
     return first_item
 
   raise ValueError(f"iterable {iterable!r} has more than one element.")
+
+
+E = TypeVar('E', bound='Enum')
+
+
+class PatternMatchError(ValueError):
+  """Issue when using pattern_match() on an enum."""
+
+
+class InexhaustivePatternsError(PatternMatchError):
+  """Not all values of the enum specified in the pattern match."""
+
+
+class UnrecognizedPatternError(PatternMatchError):
+  """A value is used that is not a part of the enum."""
+
+
+class Enum(StdLibEnum):
```

Comment:
Very hyped for this change in general. Just a bit uncertain about highjacking the `Enum` name. Makes me feel a bit uneasy as people may assume this comes from the std if they're not checking the `import` statements.
Maybe `CheckedEnum` to indicate that it's almost exactly an enum, but pattern matching is checked?

Reply:
`Enum` works for me, particularly if we add the standard library `Enum` to `build-support/bin/check_banned_imports.py` so that people _can't_ import it, which helps to avoid the confusion.

Reply:
Yes, under these conditions, it would be OK. Would also kind of enforce the safer matching, which is nice